lib/mongo/server/connectable.rb

```diff
@@ -88,10 +88,11 @@ def ensure_connected
         ensure_same_process!
         connect!
         begin
-          yield socket
-        rescue Exception => e
-          disconnect!
-          raise e
+          res = yield socket
+          success = true
+          res
+        ensure
+          success or disconnect!
```

Comment:
Sure, that's not the issue, the issue if if the `yield` doesn't raise but returns nil, we would have no way to know if `result` is nil because there was an exception, or because it is supposed to be nil.